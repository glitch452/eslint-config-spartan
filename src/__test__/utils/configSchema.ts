import { z } from 'zod';

const INITIAL_ECMA_VERSION = 3;
const INITIAL_ECMA_YEAR = 2015;
const NEXT_YEAR = new Date().getFullYear() + 1;
const MIN_ECMA_VERSION = 5;
const MAX_ECMA_VERSION = NEXT_YEAR - INITIAL_ECMA_YEAR + MIN_ECMA_VERSION + 1;

const ecmaVersionSchema = z.union([
  z.literal('latest'),
  z.literal(INITIAL_ECMA_VERSION),
  z.number().int().min(MIN_ECMA_VERSION).max(MAX_ECMA_VERSION),
  z.number().int().min(INITIAL_ECMA_YEAR).max(NEXT_YEAR),
]);
const sourceTypeSchema = z.enum(['script', 'module', 'commonjs']);
const globalConfSchema = z.boolean().or(z.enum(['off', 'readable', 'readonly', 'writable', 'writeable']));

const MIN_SEVERITY = 0;
const MAX_SEVERITY = 2;
const severitySchema = z.number().int().min(MIN_SEVERITY).max(MAX_SEVERITY);
const stringSeveritySchema = z.enum(['off', 'warn', 'error']);

const objectMetaPropertiesSchema = z.object({
  /** @deprecated Use `meta.name` instead. */
  name: z.string().optional(),
  /** @deprecated Use `meta.version` instead. */
  version: z.string().optional(),
  meta: z.object({ name: z.string().optional(), version: z.string().optional() }).optional(),
});

const esTreeParserSchema = z.union([
  objectMetaPropertiesSchema.extend({ parse: z.function() }),
  objectMetaPropertiesSchema.extend({ parseForESLint: z.function() }),
]);

const processorSchema = objectMetaPropertiesSchema.extend({
  supportsAutofix: z.boolean().optional(),
  preprocess: z.function(),
  postprocess: z.function(),
});

const parserOptionsSchema = z
  .object({
    ecmaVersion: ecmaVersionSchema,
    sourceType: sourceTypeSchema,
    ecmaFeatures: z
      .object({
        globalReturn: z.boolean(),
        impliedStrict: z.boolean(),
        jsx: z.boolean(),
        experimentalObjectRestSpread: z.boolean(),
      })
      .partial()
      .catchall(z.any()),
  })
  .partial()
  .catchall(z.any());

const languageOptionsSchema = z
  .object({
    ecmaVersion: ecmaVersionSchema,
    sourceType: sourceTypeSchema,
    globals: z.record(globalConfSchema),
    parser: esTreeParserSchema,
    parserOptions: parserOptionsSchema,
  })
  .partial();

const linterOptionsSchema = z
  .object({
    noInlineConfig: z.boolean().optional(),
    reportUnusedDisableDirectives: z.union([severitySchema, stringSeveritySchema, z.boolean()]).optional(),
  })
  .optional();

const externalSpecifierSchema = z.object({
  name: z.string().optional(),
  url: z.string().optional(),
});

const replacedBySchema = z.object({
  message: z.string().optional(),
  url: z.string().optional(),
  plugin: externalSpecifierSchema.optional(),
  rule: externalSpecifierSchema.optional(),
});

const deprecatedInfoSchema = z.object({
  message: z.string().optional(),
  url: z.string().optional(),
  replacedBy: replacedBySchema.array().optional(),
  deprecatedSince: z.string().optional(),
  availableUntil: z.string().nullable().optional(),
});

const ruleMetaDataSchema = z
  .object({
    docs: z
      .object({
        description: z.string(),
        category: z.string(),
        // Added z.string() to accommodate storybook plugin
        // Added z.record(z.unknown()) to accommodate typescript-eslint plugin
        recommended: z.union([z.boolean(), z.string(), z.record(z.unknown())]),
        url: z.string(),
      })
      .partial(),
    messages: z.record(z.string()),
    // Added z.null() added to accommodate the cypress plugin
    fixable: z.enum(['code', 'whitespace']).nullable(),
    schema: z.unknown(),
    deprecated: z.boolean().or(deprecatedInfoSchema),
    replacedBy: z.string().array().readonly(),
    // Added 'error' to accommodate the security plugin
    type: z.enum(['problem', 'suggestion', 'layout', 'error']),
    hasSuggestions: z.boolean(),
  })
  .partial();

const ruleModuleSchema = z.object({
  create: z.function(),
  meta: ruleMetaDataSchema.optional(),
});

const pluginSchema = z
  .object({
    configs: z.record(z.unknown()),
    environments: z.record(
      z.object({ globals: z.record(globalConfSchema), parserOptions: parserOptionsSchema }).partial(),
    ),
    processors: z.record(processorSchema),
    rules: z.record(ruleModuleSchema),
  })
  .partial();

export const configSchema = z
  .object({
    name: z.string(),
    files: z.array(z.string().or(z.string().array())),
    ignores: z.string().array(),
    languageOptions: languageOptionsSchema,
    linterOptions: linterOptionsSchema,
    processor: z.string().or(processorSchema),
    plugins: z.record(pluginSchema),
    rules: z.unknown(),
    settings: z.record(z.unknown()),
  })
  .partial();
