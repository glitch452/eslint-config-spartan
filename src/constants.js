export const configNamePrefix = 'eslint-config-spartan';

export const prefixes = /** @type {const} */ ({
  angular: '@angular-eslint',
  angularTemplate: '@angular-eslint/template',
  chaiExpect: 'chai-expect',
  chaiFriendly: 'chai-friendly',
  cypress: 'cypress',
  i18next: 'i18next',
  i18nJson: 'i18n-json',
  import: 'import',
  jest: 'jest',
  jestDom: 'jest-dom',
  jsDoc: 'jsdoc',
  jsxA11y: 'jsx-a11y',
  mdx: 'mdx',
  mocha: 'mocha',
  next: '@next/next',
  playwright: 'playwright',
  promise: 'promise',
  react: 'react',
  reactHooks: 'react-hooks',
  regExp: 'regexp',
  security: 'security',
  storybook: 'storybook',
  stylistic: '@stylistic',
  tailwindCss: 'tailwindcss',
  tanStackQuery: '@tanstack/query',
  testingLibrary: 'testing-library',
  typescriptEnum: 'typescript-enum',
  typescriptEsLint: '@typescript-eslint',
  unicorn: 'unicorn',
  unusedImports: 'unused-imports',
  vitest: 'vitest',
});
