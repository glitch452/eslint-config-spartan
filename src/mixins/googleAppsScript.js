import googleAppsScriptPlugin from 'eslint-plugin-googleappsscript';
import { files } from '../utils/index.js';
import { CONFIG_NAME_PREFIX } from '../constants.js';
/** @import { Linter } from 'eslint' */

/**
 * The `googleAppsScript` mixin creates an ESLint config for [eslint-plugin-googleappsscript](https://www.npmjs.com/package/eslint-plugin-googleappsscript)
 * which contains globals for the Google Apps Script environment.
 *
 * Rule Prefixes:
 *
 * - The `eslint-plugin-googleappsscript` does not contain any rules.
 * @param {object} options
 * @param {Linter.Config['files']=} options.files Set the files for this config. By default, this applies to the js/ts file extensions.
 * @returns {Linter.Config}
 */
export function googleAppsScript(options = {}) {
  return {
    name: `${CONFIG_NAME_PREFIX}/${googleAppsScript.name}`,
    files: options.files ?? [files.jsTs],
    languageOptions: {
      globals: Object.fromEntries(
        Object.entries(googleAppsScriptPlugin.environments.googleappsscript.globals).map(([k, v]) => [
          k,
          v ? 'writeable' : 'readonly',
        ]),
      ),
    },
  };
}
