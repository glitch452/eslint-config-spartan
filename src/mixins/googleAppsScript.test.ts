import { googleAppsScript } from './googleAppsScript.js';
import { configSchema } from '../__test__/utils/configSchema.js';
import { files } from '../utils/index.js';
import { CONFIG_NAME_PREFIX } from '../constants.js';

describe(googleAppsScript.name, () => {
  const config = googleAppsScript();

  it('should create a valid eslint config', () => {
    const actual = () => configSchema.parse(config);
    expect(actual).not.toThrow();
  });

  it('should configure the expected languageOptions', () => {
    expect(config.languageOptions).toMatchSnapshot();
  });

  it('should set the name value', () => {
    expect(config.name).toBe(`${CONFIG_NAME_PREFIX}/${googleAppsScript.name}`);
  });

  it('should set the default files value', () => {
    const actual = config.files;
    const expected = [files.jsTs];
    expect(actual).toStrictEqual(expected);
  });

  it('should set the given files value', () => {
    const fileValue = ['files.txt'];
    const actual = googleAppsScript({ files: fileValue }).files;
    expect(actual).toStrictEqual(fileValue);
  });

  it('should not configure any rules', () => {
    expect(config.rules).toBeUndefined();
  });
});
