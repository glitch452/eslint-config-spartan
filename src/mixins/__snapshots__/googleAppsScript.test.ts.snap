// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`googleAppsScript > should configure the expected languageOptions 1`] = `
{
  "globals": {
    "AdSense": "readonly",
    "AdminDirectory": "readonly",
    "AdminGroupsMigration": "readonly",
    "AdminGroupsSettings": "readonly",
    "AdminLicenseManager": "readonly",
    "AdminReports": "readonly",
    "AdminReseller": "readonly",
    "Analytics": "readonly",
    "AnalyticsReporting": "readonly",
    "AppsActivity": "readonly",
    "BigQuery": "readonly",
    "Browser": "readonly",
    "CacheService": "readonly",
    "Calendar": "readonly",
    "CalendarApp": "readonly",
    "CardService": "readonly",
    "Charts": "readonly",
    "Classroom": "readonly",
    "ConferenceDataService": "readonly",
    "ContactsApp": "readonly",
    "ContentService": "readonly",
    "DataStudioApp": "readonly",
    "Docs": "readonly",
    "DocumentApp": "readonly",
    "DoubleClickCampaigns": "readonly",
    "Drive": "readonly",
    "DriveActivity": "readonly",
    "DriveApp": "readonly",
    "FormApp": "readonly",
    "FusionTables": "readonly",
    "Gmail": "readonly",
    "GmailApp": "readonly",
    "GoogleAppsScript": "readonly",
    "GroupsApp": "readonly",
    "HtmlService": "readonly",
    "Jdbc": "readonly",
    "LanguageApp": "readonly",
    "LinearOptimizationService": "readonly",
    "LockService": "readonly",
    "Logger": "readonly",
    "MailApp": "readonly",
    "Maps": "readonly",
    "MimeType": "readonly",
    "Mirror": "readonly",
    "People": "readonly",
    "PropertiesService": "readonly",
    "ScriptApp": "readonly",
    "Session": "readonly",
    "Sheets": "readonly",
    "ShoppingContent": "readonly",
    "SitesApp": "readonly",
    "Slides": "readonly",
    "SlidesApp": "readonly",
    "SpreadsheetApp": "readonly",
    "TagManager": "readonly",
    "Tasks": "readonly",
    "UrlFetchApp": "readonly",
    "Utilities": "readonly",
    "XmlService": "readonly",
    "YouTube": "readonly",
    "YouTubeAnalytics": "readonly",
    "YouTubeContentId": "readonly",
    "console": "readonly",
  },
}
`;
